archivers: addition of 'lightweight' archiving for BagIt

krzysztof opened the pull request:
* Adds a functionality for BagIt archiver to write only
  the new files to the archive location, and refer to the
  already archived content in fetch.txt files.

Signed-off-by: Krzysztof Nowak <k.nowak@cern.ch>

slint commented on invenio_sipstore/archivers/bagit_archiver.py:
```diff
@@ -54,25 +63,137 @@ def get_all_files(self):
+        When specifying 'patch_of' parameter the 'include_missing_files'
+        flag determines whether files that are missing in the archived SIP
+        (w.r.t. the SIP specified in 'patch_of') should be treated as
+        explicitly deleted (include_missing_files=False) or of all
```
...or **if** all

slint commented on invenio_sipstore/archivers/bagit_archiver.py:
```diff
@@ -54,25 +63,137 @@ def get_all_files(self):
         """
+        sipfiles = self.sip.files
+        if patch_of:
+            fetch_info = {}
```
Throughout the next operations it's a bit unclear what's the general format/contents of the `fetch_info` dict. Maybe it's worth giving a brief "schema" definition or example here (or referencing somehow the output format from `self._copy_files` I guess...)

slint commented on invenio_sipstore/archivers/bagit_archiver.py:
```diff
@@ -54,25 +63,137 @@ def get_all_files(self):
 
+        if create_bagit_metadata:
+            # Store what we archived in BagIt
+            bagit_meta = {}
```
Same thing as with the `fetch_info` dict, some brief example or definition of the "schema" definition would help.

krzysztof replied:
Added the schema https://github.com/inveniosoftware/invenio-sipstore/pull/30/files#diff-a56fed499f498717abb66ba8c12807f9

slint commented on invenio_sipstore/archivers/bagit_archiver.py:
```diff
@@ -82,8 +203,14 @@ def autogenerate_tags(self, files_info):
 
+    def get_fetch_file(self, files_info):
+        """Generate the contents of the fetch.txt file."""
+        content = ('{0} {1} {2}'.format(c['path'], c['size'], f)
```
I believe the FILENAME (and maybe even URL) as well must be url-encoded to avoid issues with spaces:

> If FILENAME begins with a slash character, the destination MUST still be treated as relative to the bag base directory.  One or more linear whitespace characters (spaces or tabs) MUST separate these three values, and any such characters in the URL MUST be percent-encoded [RFC3986].

krzysztof replied:
Resolved.

krzysztof commented:
I have added a jsonschema for the bagit-archiver: https://github.com/inveniosoftware/invenio-sipstore/pull/30/files#diff-a56fed499f498717abb66ba8c12807f9

It's sadly not registered so just like in agent into, it's not yet added to the created objects's JSONs.

@remileduc At the moment the "create" method of the base archiver is returning a dict like:
```
{
    'data/filename.txt': {
       'size':12,
       'checksum': '...',
       'path': '/abs/path/to/disk',
       'file_uuid':'<FileInstance UUID>'
    },
   'data/filename2.txt': {
     (...)
   }
}
```

This is because _copy_files and _save_file return it in this structure.

Would you mind if I refactored this (and all archiver's code) to return a list with a flat-structure?
Like this:
```
[
  {
     'filename': 'data/filename.txt'
     'size':12,
     'checksum': '...',
     'path': '/abs/path/to/disk',
     'file_uuid':'FileInstance UUID if relevant'}
  },
  (...)
]
```

remileduc commented:
sounds fine for me :)
